StatusRuntimeException from inside submit not propogated

Discovered as part of https://github.com/deephaven/deephaven-core/pull/1787, `StatusRuntimeException`s that are thrown during execution of `io.deephaven.server.session.SessionState.ExportBuilder#submit(java.util.concurrent.Callable<T>)` don't have their message propogated to the client.

This is somewhat related to #801. In general, we wanted to avoid leaking internal error information to malicious users.

Note that ExportObjects do not require the use of gRPC. I don't believe that `submit` can throw a StatusRuntimeException; that lambda is deferred to a thread pool. There is a decoupling between errors that occur when the work is processed. These errors are supposed to be passed to the export object's error handler. In gRPCs case these are supposed to pass through StatusRuntimeExceptions to the observer, and other exception types are supposed to be wrapped with the UUID identifier.

I looked at a few items in the follow up list to try and create a repro. I uploaded this to https://github.com/nbauernfeind/deephaven-core/tree/gh1871_repro. Server side gets this error:
```
Internal Error '63ee9a27-e53a-494d-ade3-9ae6771dd200' io.grpc.StatusRuntimeException: INVALID_ARGUMENT: Table can't be used as an input table
        at io.grpc.Status.asRuntimeException(Status.java:535)
        at io.grpc.protobuf.StatusProto.toStatusRuntimeException(StatusProto.java:52)
        at io.deephaven.proto.util.Exceptions.statusRuntimeException(Exceptions.java:14)
        at io.deephaven.extensions.barrage.util.GrpcUtil.statusRuntimeException(GrpcUtil.java:95)
        at io.deephaven.server.table.inputtables.InputTableServiceGrpcImpl.lambda$addTableToInputTable$1(InputTableServiceGrpcImpl.java:56)
        at io.deephaven.server.session.SessionState$ExportBuilder.lambda$submit$2(SessionState.java:1308)
        at io.deephaven.server.session.SessionState$ExportObject.doExport(SessionState.java:856)
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
        at java.util.concurrent.FutureTask.run(FutureTask.java:264)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1130)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:630)
        at io.deephaven.server.runner.DeephavenApiServerModule$ThreadFactory.lambda$newThread$0(DeephavenApiServerModule.java:162)
        at java.lang.Thread.run(Thread.java:832)
```

The java client gets:
```
Caused by: io.grpc.StatusRuntimeException: FAILED_PRECONDITION: Details Logged w/ID '63ee9a27-e53a-494d-ade3-9ae6771dd200'
	at io.grpc.Status.asRuntimeException(Status.java:535)
	at io.grpc.stub.ClientCalls$StreamObserverToCallListenerAdapter.onClose(ClientCalls.java:487)
	at io.grpc.internal.ClientCallImpl.closeObserver(ClientCallImpl.java:562)
	at io.grpc.internal.ClientCallImpl.access$300(ClientCallImpl.java:70)
	at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl$1StreamClosed.runInternal(ClientCallImpl.java:743)
	at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl$1StreamClosed.runInContext(ClientCallImpl.java:722)
	at io.grpc.internal.ContextRunnable.run(ContextRunnable.java:37)
	at io.grpc.internal.SerializingExecutor.run(SerializingExecutor.java:133)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1130)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:630)
	at java.base/java.lang.Thread.run(Thread.java:832)
```

One gotcha here is that the batch API cannot send multiple status results for the single rpc call. The batch's per export response only allows for an error info string.  It seems fair to fail with the first detected error and keep the UUIDs for batch; one shot RPCs should propagate that status runtime exception.

--- a/server/src/main/java/io/deephaven/server/session/SessionState.java
+++ b/server/src/main/java/io/deephaven/server/session/SessionState.java
@@ -524,6 +524,7 @@ public class SessionState {
         /** used to identify and propagate error details */
         private String errorId;
         private String dependentHandle;
+        private Exception caughtException;
 
         /**
          * @param exportId the export id for this export
@@ -740,7 +741,7 @@ public class SessionState {
                 if (errorId == null) {
                     assignErrorId();
                 }
-                safelyExecute(() -> errorHandler.onError(state, errorId, dependentHandle));
+                safelyExecute(() -> errorHandler.onError(state, errorId, caughtException, dependentHandle));
             }
 
             if (state == ExportNotification.State.EXPORTED || isExportStateTerminal(state)) {
@@ -750,6 +751,7 @@ public class SessionState {
                 parents = Collections.emptyList();
                 exportMain = null;
                 errorHandler = null;
+                caughtException = null;
             }
 
             if ((state == ExportNotification.State.EXPORTED && isNonExport()) || isExportStateTerminal(state)) {
@@ -841,7 +843,6 @@ public class SessionState {
                 }
                 setState(ExportNotification.State.RUNNING);
             }
-            Exception exception = null;
             boolean shouldLog = false;
             int evaluationNumber = -1;
             QueryProcessingResults queryProcessingResults = null;
@@ -858,7 +859,7 @@ public class SessionState {
                     shouldLog = QueryPerformanceRecorder.getInstance().endQuery();
                 }
             } catch (final Exception err) {
-                exception = err;
+                caughtException = err;
                 synchronized (this) {
                     if (!isExportStateTerminal(state)) {
                         assignErrorId();
@@ -867,12 +868,12 @@ public class SessionState {
                     }
                 }
             } finally {
-                if (exception != null && queryProcessingResults != null) {
-                    queryProcessingResults.setException(exception.toString());
+                if (caughtException != null && queryProcessingResults != null) {
+                    queryProcessingResults.setException(caughtException.toString());
                 }
                 QueryPerformanceRecorder.resetInstance();
             }
-            if ((shouldLog || exception != null) && queryProcessingResults != null) {
+            if ((shouldLog || caughtException != null) && queryProcessingResults != null) {
                 final MemoryTableLoggers memLoggers = MemoryTableLoggers.getInstance();
                 final QueryPerformanceLogLogger qplLogger = memLoggers.getQplLogger();
                 final QueryOperationPerformanceLogLogger qoplLogger = memLoggers.getQoplLogger();
@@ -1141,7 +1142,9 @@ public class SessionState {
          * @param dependentExportId an identifier for the export id of the dependent that caused the failure if
          *        applicable
          */
-        void onError(final ExportNotification.State resultState, @Nullable final String errorContext,
+        void onError(final ExportNotification.State resultState,
+                final String errorContext,
+                @Nullable final Exception cause,
                 @Nullable final String dependentExportId);
     }
     @FunctionalInterface
@@ -1248,7 +1251,12 @@ public class SessionState {
          * @return this builder
          */
         public ExportBuilder<T> onErrorHandler(final ExportErrorGrpcHandler errorHandler) {
-            return onError(((resultState, errorContext, dependentExportId) -> {
+            return onError(((resultState, errorContext, cause, dependentExportId) -> {
+                if (cause instanceof StatusRuntimeException) {
+                    errorHandler.onError((StatusRuntimeException) cause);
+                    return;
+                }
+
                 final String dependentStr = dependentExportId == null ? ""
                         : (" (related parent export id: " + dependentExportId + ")");
                 errorHandler.onError(GrpcUtil.statusRuntimeException(

--- a/server/src/main/java/io/deephaven/server/session/SessionState.java
+++ b/server/src/main/java/io/deephaven/server/session/SessionState.java
@@ -751,7 +751,6 @@ public class SessionState {
                 parents = Collections.emptyList();
                 exportMain = null;
                 errorHandler = null;
-                caughtException = null;
             }
 
             if ((state == ExportNotification.State.EXPORTED && isNonExport()) || isExportStateTerminal(state)) {
@@ -774,6 +773,9 @@ public class SessionState {
             if (parent != null && isExportStateTerminal(parent.state)) {
                 synchronized (this) {
                     errorId = parent.errorId;
+                    if (parent.caughtException instanceof StatusRuntimeException) {
+                        caughtException = parent.caughtException;
+                    }
                     ExportNotification.State terminalState = ExportNotification.State.DEPENDENCY_FAILED;
 
                     if (errorId == null) {
@@ -976,6 +978,7 @@ public class SessionState {
         protected synchronized void destroy() {
             super.destroy();
             result = null;
+            caughtException = null;
         }
 
         /**

--- a/server/src/main/java/io/deephaven/server/table/TableServiceGrpcImpl.java
+++ b/server/src/main/java/io/deephaven/server/table/TableServiceGrpcImpl.java
@@ -47,6 +47,8 @@ import io.deephaven.server.session.SessionState;
 import io.deephaven.server.session.SessionState.ExportBuilder;
 import io.deephaven.server.session.TicketRouter;
 import io.deephaven.server.table.ops.GrpcTableOperation;
+import io.grpc.Status;
+import io.grpc.StatusRuntimeException;
 import io.grpc.stub.ServerCallStreamObserver;
 import io.grpc.stub.StreamObserver;
 
@@ -55,6 +57,7 @@ import java.util.List;
 import java.util.Map;
 import java.util.Objects;
 import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicReference;
 import java.util.stream.Collectors;
 
 import static io.deephaven.extensions.barrage.util.GrpcUtil.safelyExecute;
@@ -291,6 +294,20 @@ public class TableServiceGrpcImpl extends TableServiceGrpc.TableServiceImplBase 
 
             // step 4: submit the batched operations
             final AtomicInteger remaining = new AtomicInteger(exportBuilders.size());
+            final AtomicReference<StatusRuntimeException> firstFailure = new AtomicReference<>();
+
+            final Runnable onOneResolved = () -> {
+                if (remaining.decrementAndGet() == 0) {
+                    safelyExecuteLocked(responseObserver, () -> {
+                        final StatusRuntimeException failure = firstFailure.get();
+                        if (failure != null) {
+                            responseObserver.onError(failure);
+                        } else {
+                            responseObserver.onCompleted();
+                        }
+                    });
+                }
+            };
 
             for (int i = 0; i < exportBuilders.size(); ++i) {
                 final BatchExportBuilder<?> exportBuilder = exportBuilders.get(i);
@@ -303,25 +320,29 @@ public class TableServiceGrpcImpl extends TableServiceGrpc.TableServiceImplBase 
                     resultId = ExportTicketHelper.tableReference(exportId);
                 }
 
-                exportBuilder.exportBuilder.onError((result, errorContext, dependentId) -> {
+                exportBuilder.exportBuilder.onError((result, errorContext, cause, dependentId) -> {
+                    String errorInfo = errorContext;
+                    if (dependentId != null) {
+                        errorInfo += " dependency: " + dependentId;
+                    }
+                    if (cause instanceof StatusRuntimeException) {
+                        errorInfo += " cause: " + cause.getMessage();
+                        firstFailure.compareAndSet(null, (StatusRuntimeException) cause);
+                    }
+                    final String finalErrorInfo = errorInfo;
                     safelyExecuteLocked(responseObserver,
                             () -> responseObserver.onNext(ExportedTableCreationResponse.newBuilder()
                                     .setResultId(resultId)
                                     .setSuccess(false)
-                                    .setErrorInfo(errorContext + " dependency: " + dependentId)
+                                    .setErrorInfo(finalErrorInfo)
                                     .build()));
-
-                    if (remaining.decrementAndGet() == 0) {
-                        safelyExecuteLocked(responseObserver, responseObserver::onCompleted);
-                    }
+                    onOneResolved.run();
                 }).submit(() -> {
                     final Table table = exportBuilder.doExport();
 
                     safelyExecuteLocked(responseObserver,
                             () -> responseObserver.onNext(ExportUtil.buildTableCreationResponse(resultId, table)));
-                    if (remaining.decrementAndGet() == 0) {
-                        safelyExecuteLocked(responseObserver, responseObserver::onCompleted);
-                    }
+                    onOneResolved.run();
 
                     return table;
                 });

--- a/server/src/test/java/io/deephaven/server/session/SessionStateTest.java
+++ b/server/src/test/java/io/deephaven/server/session/SessionStateTest.java
@@ -7,6 +7,7 @@ import io.deephaven.base.verify.Assert;
 import io.deephaven.base.verify.AssertionFailure;
 import io.deephaven.engine.util.NoLanguageDeephavenSession;
 import io.deephaven.proto.util.ExportTicketHelper;
+import io.deephaven.server.table.ExportTableUpdateListenerTest;
 import io.deephaven.time.DateTimeUtils;
 import io.deephaven.engine.liveness.LivenessArtifact;
 import io.deephaven.engine.liveness.LivenessReferent;
@@ -18,6 +19,7 @@ import io.deephaven.proto.backplane.grpc.ExportNotification;
 import io.deephaven.proto.backplane.grpc.Ticket;
 import io.deephaven.util.SafeCloseable;
 import io.deephaven.util.auth.AuthContext;
+import io.grpc.Status;
 import io.grpc.StatusRuntimeException;
 import io.grpc.stub.StreamObserver;
 import org.apache.commons.lang3.mutable.MutableBoolean;
@@ -1246,6 +1248,44 @@ public class SessionStateTest {
         });
         scheduler.runUntilQueueEmpty();
         Assert.eq(clr.refCount, "clr.refCount", 0);
+    }
+
+    @Test
+    public void testCascadingStatusRuntimeFailureDeliversToErrorHandler() {
+        final SessionState.ExportObject<Object> e1 = session.newExport(nextExportId++)
+                .submit(() -> {
+                    throw Status.DATA_LOSS.asRuntimeException();
+                });
+
+        final MutableBoolean submitRan = new MutableBoolean();
+        final MutableObject<Throwable> caughtErr = new MutableObject<>();
+        final StreamObserver<?> observer = new StreamObserver<>() {
+            @Override
+            public void onNext(Object value) {
+                throw new RuntimeException("this should not reach test framework");
+            }
+
+            @Override
+            public void onError(Throwable t) {
+                caughtErr.setValue(t);
+            }
+
+            @Override
+            public void onCompleted() {
+                throw new RuntimeException("this should not reach test framework");
+            }
+        };
+        session.newExport(nextExportId++)
+                .onError(observer)
+                .require(e1)
+                .submit(submitRan::setTrue);
+
+        scheduler.runUntilQueueEmpty();
+        Assert.eqFalse(submitRan.booleanValue(), "submitRan.booleanValue()");
+        Assert.eqTrue(caughtErr.getValue() instanceof StatusRuntimeException, "caughtErr.getValue()");
+
+        final StatusRuntimeException sre = (StatusRuntimeException) caughtErr.getValue();
+        Assert.eq(sre.getStatus(), "sre.getStatus()", Status.DATA_LOSS, "Status.DATA_LOSS");
     }
 
     private static long getExportId(final ExportNotification notification) {

